src/FSharp.Data.GraphQL.Shared/Validation.fs

```diff
@@ -178,4 +193,199 @@ module Ast =
+                else acc) acc
+        | InlineFragment frag -> frag.SelectionSet |> List.map (checkArgumentUniqueness acc fragmentDefinitions) |> List.reduce (@)
+        | FragmentSpread spread ->
+            match fragmentDefinitions |> List.tryFind (fun f -> f.Name.IsSome && f.Name.Value = spread.Name) with
+            | Some frag -> frag.SelectionSet |> List.map (checkArgumentUniqueness acc fragmentDefinitions) |> List.reduce (@)
+            | None -> acc
+
+    let validateArgumentUniqueness (ast : Document) =
+        let fragmentDefinitions = getFragmentDefinitions ast
+        ast.Definitions
+        |> List.collect (function
+            | FragmentDefinition frag -> frag.SelectionSet
+            | OperationDefinition op -> op.SelectionSet)
+        |> List.fold (fun acc def -> checkArgumentUniqueness acc fragmentDefinitions def) Success
+
+    let rec private checkRequiredArguments (acc : ValidationResult) (schemaInfo : SchemaInfo) (selection : SelectionInfo) =
+        let acc =
+            selection.ArgDefs
+            |> Option.map (Array.fold (fun acc argDef ->
+                match argDef.Type.Kind with
+                | TypeKind.NON_NULL when argDef.DefaultValue.IsNone ->
+                    match selection.Arguments |> List.tryFind (fun arg -> arg.Name = argDef.Name) with
+                    | Some arg when arg.Value <> EnumValue "null" -> acc // TODO: null values are being mapped into an enum value, this should be fixed in the parser! A null value must be able to be parsed.
+                    | _ -> acc @ Error [ sprintf "Argument '%s' of field '%s' of type '%s' is required and does not have a default value." argDef.Name selection.Name selection.ParentType.Name ]
+                | _ -> acc) acc)
+            |> Option.defaultValue acc
+        selection.Directives
+        |> List.fold (fun acc directive ->
+            match schemaInfo.Directives |> Array.tryFind (fun d -> d.Name = directive.Name) with
+            | Some directiveType ->
+                directiveType.Args
+                |> Array.fold (fun acc argDef ->
+                    match argDef.Type.Kind with
+                    | TypeKind.NON_NULL when argDef.DefaultValue.IsNone ->
+                        match selection.Arguments |> List.tryFind (fun arg -> arg.Name = argDef.Name) with
+                        | Some arg when arg.Value <> EnumValue "null" -> acc
+                        | _ -> acc @ Error [ sprintf "Argument '%s' of directive '%s' of field '%s' of type '%s' is required and does not have a default value." argDef.Name directiveType.Name selection.Name selection.ParentType.Name ]
+                    | _ -> acc) acc
+            | None -> acc) acc
+
+    let validateRequiredArguments (schemaInfo : SchemaInfo) (ast : Document) =
+        let fragmentDefinitions = getFragmentDefinitions ast
+        ast.Definitions
+        |> List.choose (tryGetSelectionSetAndType schemaInfo)
```

Comment:
It might make sense to define a fused version of `List.choose` followed by `List.fold`

Reply:
The code did change for this part, and I made it only use `List.fold`.